File: tests/src/com.google.protobuf/protobuf-javalite/3.25.3/src/test/java/com_google_protobuf/protobuf_javalite/ExtensionSchemasTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package com_google_protobuf.protobuf_javalite;

import static java.lang.invoke.MethodType.methodType;
import static org.assertj.core.api.Assertions.assertThat;

import java.lang.invoke.MethodHandle;
import java.lang.invoke.MethodHandles;
import org.junit.jupiter.api.Test;

public class ExtensionSchemasTest {
    private static final String MESSAGE_OPTIONS_CLASS_NAME =
            "com.google.protobuf.DescriptorProtos$MessageOptions";
    private static final String PROTOBUF_CLASS_NAME = "com.google.protobuf.Protobuf";
    private static final String SCHEMA_CLASS_NAME = "com.google.protobuf.Schema";

    @Test
    void createsFullRuntimeExtensionSchemaForExtendableGeneratedMessage() throws Throwable {
        Class<?> messageOptionsClass = Class.forName(MESSAGE_OPTIONS_CLASS_NAME);
        Class<?> protobufClass = Class.forName(
                PROTOBUF_CLASS_NAME,
                true,
                messageOptionsClass.getClassLoader()
        );
        Class<?> schemaClass = Class.forName(
                SCHEMA_CLASS_NAME,
                true,
                protobufClass.getClassLoader()
        );
        MethodHandles.Lookup lookup = MethodHandles.privateLookupIn(
                protobufClass,
                MethodHandles.lookup()
        );
        MethodHandle getInstance = lookup.findStatic(
                protobufClass,
                "getInstance",
                methodType(protobufClass)
        );
        MethodHandle schemaFor = lookup.findVirtual(
                protobufClass,
                "schemaFor",
                methodType(schemaClass, Class.class)
        );

        Object protobuf = getInstance.invoke();
        Object schema = schemaFor.invoke(protobuf, messageOptionsClass);

        assertThat(schema).isNotNull();
    }
}

File: tests/src/org.apache.pekko/pekko-protobuf-v3_2.13/1.2.0/src/test/scala/org_apache_pekko/pekko_protobuf_v3_2_13/ExtensionSchemasTest.scala
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_apache_pekko.pekko_protobuf_v3_2_13

import java.lang.invoke.MethodHandle
import java.lang.invoke.MethodHandles
import java.lang.invoke.MethodType.methodType

import org.apache.pekko.protobufv3.internal.DescriptorProtos
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class ExtensionSchemasTest {
  @Test
  def fullRuntimeExtensionSchemaIsLoadedForProto2Messages(): Unit = {
    val message: ExtensionSchemasProto2Message = ExtensionSchemasProto2Message.getDefaultInstance

    message.parseEmptyInputThroughSchema()

    assertThat(message.getDescriptorForType.getFile.toProto.getSyntax).isEqualTo("proto2")
  }

  @Test
  def fullRuntimeSchemaFactoryLoadsExtensionSchemaFullConstructor(): Unit = {
    val protobufClass: Class[?] = Class.forName("org.apache.pekko.protobufv3.internal.Protobuf")
    val schemaClass: Class[?] = Class.forName("org.apache.pekko.protobufv3.internal.Schema")
    val lookup: MethodHandles.Lookup = MethodHandles.privateLookupIn(
      protobufClass,
      MethodHandles.lookup()
    )
    val getInstance: MethodHandle = lookup.findStatic(
      protobufClass,
      "getInstance",
      methodType(protobufClass)
    )
    val schemaFor: MethodHandle = lookup.findVirtual(
      protobufClass,
      "schemaFor",
      methodType(schemaClass, classOf[Class[_]])
    )
    val protobuf: Object = getInstance.invokeWithArguments()
    val schema: Object = schemaFor.invokeWithArguments(
      protobuf,
      classOf[DescriptorProtos.FileDescriptorSet]
    )

    assertThat(schema).isNotNull
  }
}
